### PR TITLE
fix(node/zlib): accept dataview and buffer in zlib bindings

### LINKED ISSUE #20516: node:zlib module has numerous bugs, inconsistencies, and errors
I've noticed quite a few issues with the node:zlib implementation. Wasn't sure if I should open them up in individual issues, but decided to just do it unified since they're all related to the same module.

First off, I'd like to reference the [official Node.js documentation on their zlib module](https://nodejs.org/api/zlib.html#zlibbrotlicompresssyncbuffer-options). **Note** that the accepted input types by **_all_** convenience methods is `Buffer | TypedArray | DataView | ArrayBuffer | string`.

1.  ~~brotli compression fails on data > 4096B~~ (fixed in #20301)

2. `zlib.brotliCompressSync` only accepts inputs that are Buffer, TypedArray, or string. It accepts all types of TypedArray, it seems, but DataViews and ArrayBuffers cause it to throw a vague `Failed to compress` exception. Inconsistent with the expected behavior above. See below for a screenshot.
Screenshot 2023-09-15 at 8 54 16 AM

3. `zlib.brotliCompress` returns what appears to be a native Uint8Array, rather than a Node.js Buffer like its synchronous counterpart. Definitely inconsistent behavior, and it gets more inconsistent when trying different input types. Any type other than a Buffer, string, or Uint8Array throws an error saying it "was not a typed ArrayBufferView".... which is not true. See below.
Screenshot 2023-09-15 at 8 55 41 AM

4. The non-brotli convenience functions all seem to throw their own errors that have no discernable pattern or logic when it comes to which binary types should be allowed and which should be banned. Here's another screenshot showing gzip and gzipSync's erratic behavior:
Screenshot 2023-09-15 at 9 03 28 AM

I'd spend more time and document the behavior on each of the remaining methods but unfortunately I'm out of time right now. I'll try to follow this up with additional info and (hopefully) a PR, if I'm able to squeeze in a fix for this tonight/tomorrow. 

Thanks!

### PR DESCRIPTION
**Pull request by JoviDeCroock**
fixes #20516 
Follow up to #21747 and #21746

This tackles the last point of #20516 where certain inputs weren't accepted in the other zlib methods

This adds the `toU8` conversion of `_brotli` to `_zlib.mjs`, when we create the ZLibBuffer, we'll sanitize the input. I noticed that the async had no handler for `string` input so I added that as well.

Not sure what is going wrong here for the parallel test 😅

### CODE REVIEW

**Comment by bartlomieju** [comment]
@littledivy please review

**Review by littledivy — APPROVED** [review]
LGTM!

**Comment by littledivy** [comment]
@JoviDeCroock The following Node compat test is failing:

```
Node.js compatibility ... Node.js compatibility "parallel/test-zlib-truncated.js" => ./cli/tests/node_compat/test.ts:58:11
error: AssertionError: Failed assertion: "parallel/test-zlib-truncated.js" failed:

error: Uncaught AssertionError: Missing expected exception.
    at new AssertionError (ext:deno_node/assertion_error.ts:414:11)
    at Function.throws (node:assert:133:11)
    at file:///home/runner/work/deno/deno/cli/tests/node_compat/test/parallel/test-zlib-truncated.js:49:12
    at Gzip.onEnd (ext:deno_node/_zlib.mjs:212:5)
    at Gzip.emit (ext:deno_node/_stream.mjs:1859:11)
    at endReadableNT (ext:deno_node/_stream.mjs:3648:16)
    at Array.processTicksAndRejections (ext:deno_node/_next_tick.ts:33:15)
    at eventLoopTick (ext:core/01_core.js:180:29)
```

**Comment by JoviDeCroock** [comment]
@littledivy Yes I can see that 😅 as mentioned in the description I can't quite tell why or what is failing as the whole failure is very vaguely worded

File: ext/node/polyfills/_zlib.mjs [hunk 1]
```diff
@@ -155,10 +155,27 @@ export const inflateRawSync = function (buffer, opts) {
...
+  if (typeof input === "string") input = Buffer.from(input);
+
+  if (
+    !Buffer.isBuffer(input) &&
+    (input.buffer && !input.buffer.constructor === ArrayBuffer)
+  ) throw new TypeError("Not a string, buffer or dataview");
+
+  if (input.buffer) {
+    input = new Uint8Array(input.buffer);
+  }
```
Comment (littledivy) [inline ext/node/polyfills/_zlib.mjs:168]:
```suggestion
  if (input.buffer) {
    input = new Uint8Array(input.buffer);
  }
```

I'm not sure if this is the reason for test failure but the input view should respect offsets.

```js
input = new Uint8Array(input.buffer, input.byteOffset, input.byteLength)
```

In the test, the compressed buffer is truncated and passed to decompress(). It expects decompress() to fail.
https://github.com/denoland/deno/blob/59cfe20f3be727096d328ecb2196478ae01447f6/cli/tests/node_compat/test/parallel/test-zlib-truncated.js#L48-L51

**Pull request merged**